internal/rollout/device_selection/batch_sequence_selector.go

```diff
@@ -0,0 +1,520 @@
+}
+
+// A group of device is considered as completed successfully if the rendered template version is the same as the
+// template version of the fleet and same-rendered-version is true
+func (b *batchSelection) isUpdateCompletedSuccessfully(c store.CompletionCount) bool {
+	return c.SameTemplateVersion && c.SameRenderedVersion
+}
+
+// IsComplete checks is the total number of devices in a batch is the same as the number of completed
+func (b *batchSelection) IsComplete(ctx context.Context) (bool, error) {
+	counts, err := b.store.Device().CompletionCounts(ctx, b.orgId, util.ResourceOwner(api.FleetKind, b.fleetName), b.templateVersionName, &b.updateTimeout)
+	if err != nil {
+		return false, err
+	}
+
+	// A device is counted in total if it has completed successfully, or it is in update.
+	total := lo.Sum(lo.Map(counts, func(c store.CompletionCount, _ int) int64 {
+		return c.Count
+	}))
+
+	// A device is counted as completed if it has completed successfully or, it is in error state or its update is timed out
+	complete := lo.Sum(lo.Map(counts, func(c store.CompletionCount, _ int) int64 {
+		return lo.Ternary(b.isUpdateCompletedSuccessfully(c) || c.SameTemplateVersion && (c.UpdatingReason == api.UpdateStateError || c.UpdateTimedOut), c.Count, 0)
+	}))
+	return total == complete, nil
+}
+
+func (b *batchSelection) SetSuccessPercentage(ctx context.Context) error {
+	counts, err := b.store.Device().CompletionCounts(ctx, b.orgId, util.ResourceOwner(api.FleetKind, b.fleetName), b.templateVersionName, nil)
+	if err != nil {
+		return err
+	}
+
+	// A device is counted in total if it has completed successfully, or it is not disconnected (Unknown), or it is in update.
+	total := lo.Sum(lo.Map(counts, func(c store.CompletionCount, _ int) int64 {
+		return c.Count
+	}))
+	successful := lo.Sum(lo.Map(counts, func(c store.CompletionCount, _ int) int64 {
+		return lo.Ternary(b.isUpdateCompletedSuccessfully(c), c.Count, 0)
+	}))
+	if total == 0 {
+		return nil
+	}
+	successPercentage := successful * 100 / total
+	b.log.Infof("%v/%s:In SetSuccessPercentage: %d", b.orgId, b.fleetName, successPercentage)
+	annotations := map[string]string{
+		api.FleetAnnotationLastBatchSuccessPercentage: strconv.FormatInt(successPercentage, 10),
+	}
+	return b.store.Fleet().UpdateAnnotations(ctx, b.orgId, b.fleetName, annotations, nil)
+}
+
+func (b *batchSelection) batchCounts(ctx context.Context) (int, int, error) {
+	parts := newQuerySelectorParts().
+		withOwner(b.fleetName).
+		withLabelSelector(b.batch.Selector)
+	total, err := b.store.Device().Count(ctx, b.orgId, parts.listParams())
+	if err != nil {
+		return 0, 0, err
+	}
+	rolledOut, err := b.store.Device().Count(ctx, b.orgId, parts.withRolledOut(b.templateVersionName).listParams())
+	if err != nil {
+		return 0, 0, err
+	}
+	return int(rolledOut), int(total), nil
+}
+
+func (b *batchSelection) calculateLimit(ctx context.Context) (*int, error) {
+	if b.batch == nil || b.batch.Limit == nil {
+		return nil, nil
```

Comment:
nit: if I am a consumer of this method the general expectation is I should not need to nil check the returned value if the error is non nil. In other words it would be unexpected to panic using the value returned here.

Maybe return (int, error)?

Reply:
In case a limit doesn't exist then nil is returned.  It means that limit should not be applied.